contrib/storage-ipfs/src/main/java/org/apache/drill/exec/store/ipfs/IPFSHelper.java

```diff
@@ -0,0 +1,286 @@
+   * @param in the parameter for op
+   * @param timeout consider the execution has timed out after this amount of time in seconds
+   * @param <T>
+   * @param <R>
+   * @param <E>
+   * @return R the result of the operation
+   * @throws E
+   */
+  public <T, R, E extends Exception> R timedFailure(ThrowingFunction<T, R, E> op, T in, int timeout) throws E {
+    Callable<R> task = () -> op.apply(in);
+    return timedFailure(task, timeout, TimeUnit.SECONDS);
+  }
+
+  public <R, E extends Exception> R timedFailure(ThrowingSupplier<R, E> op, int timeout) throws E {
+    Callable<R> task = op::get;
+    return timedFailure(task, timeout, TimeUnit.SECONDS);
+  }
+
+  private <R, E extends Exception> R timedFailure(Callable<R> task, int timeout, TimeUnit timeUnit) throws E {
+    Future<R> res = executorService.get().submit(task);
+    try {
+      return res.get(timeout, timeUnit);
+    } catch (ExecutionException e) {
+      throw (E) e.getCause();
+    } catch (TimeoutException e) {
+      throw UserException.executionError(e).message("IPFS operation timed out").build(logger);
+    } catch (CancellationException | InterruptedException e) {
+      throw UserException.executionError(e).build(logger);
+    }
+  }
+
+  public static Optional<String> pickPeerHost(List<MultiAddress> peerAddrs) {
+    String localAddr = null;
+    for (MultiAddress addr : peerAddrs) {
+      String host = addr.getHost();
+      try {
+        InetAddress inetAddress = InetAddress.getByName(host);
+        if (inetAddress.isLoopbackAddress()) {
+          continue;
+        }
+        if (inetAddress.isSiteLocalAddress() || inetAddress.isLinkLocalAddress()) {
+          //FIXME we don't know which local address can be reached; maybe check with InetAddress.isReachable?
+          localAddr = host;
+        } else {
+          return Optional.of(host);
+        }
+      } catch (UnknownHostException e) {
+        continue;
+      }
+    }
+
+    return Optional.ofNullable(localAddr);
+  }
+
+  public Optional<String> getPeerDrillHostname(Multihash peerId) {
+    return getPeerData(peerId, "drill-hostname").map(Strings::fromByteArray);
+  }
+
+  public boolean isDrillReady(Multihash peerId) {
+    try {
+      return getPeerData(peerId, "drill-ready").isPresent();
+    } catch (RuntimeException e) {
+      return false;
+    }
+  }
+
+  public Optional<Multihash> getIPNSDataHash(Multihash peerId) {
+    Optional<List<MerkleNode>> links = getPeerLinks(peerId);
+    if (!links.isPresent()) {
+      return Optional.empty();
+    }
+
+    return links.get().stream()
+        .filter(l -> l.name.equals(Optional.of("drill-data")))
+        .findFirst()
+        .map(l -> l.hash);
+  }
+
+
+  private Optional<byte[]> getPeerData(Multihash peerId, String key) {
+    Optional<List<MerkleNode>> links = getPeerLinks(peerId);
+    if (!links.isPresent()) {
+      return Optional.empty();
+    }
+
+    return links.get().stream()
```

Comment:
Is streaming really necessary here, comparing to ordinary `for` loop?

Reply:
Changed in ca71f95.